packages/hyperion-flowlet/src/FlowletManager.ts

```diff
@@ -6,15 +6,60 @@ import { Hook } from "@hyperion/hook";
 import { getFunctionInterceptor } from "@hyperion/hyperion-core/src/FunctionInterceptor";
 import { CallbackType, interceptEventListener, isEventListenerObject } from "@hyperion/hyperion-dom/src/IEventListener";
 import { Flowlet } from "./Flowlet";
-
+import { assert } from "@hyperion/global";
+import { TimedTrigger } from "@hyperion/hyperion-util/src/TimedTrigger";
 
 const IS_FLOWLET_SETUP_PROP_NAME = `__isFlowletSetup`;
 
+const CLEANUP_PERIOD = 3000; // ms call cleanup at most every 3 seconds
+const MIN_STACK_SIZE_FOR_QUICK_CLEANUP = 10; // clean up pending tasks if system was not idle and stack size is greater than this value
+
+
 export class FlowletManager<T extends Flowlet = Flowlet> {
   private _flowletStack: T[] = [];
   private _top: T | null = null; // To optimize for faster reading of top;
 
-  constructor(public flowletCtor: new (flowletName: string, parent?: T | null) => T) { }
+  /**
+   * If we don't succeed at poping the top of stack, we keep a list of those
+   * iterms and periodically remove them to ensure the size of the stack does
+   * not keep growing.
+   */
+  private _pendingPops = new Set<T>();
+
+  constructor(public flowletCtor: new (flowletName: string, parent?: T | null) => T) {
+    if (__DEV__) {
+      this.onPush.add(flowlet => {
+        assert(flowlet != null, `Unexpected NULL flowlet value pushed to stack!`);
+      });
+    }
+    this.scheduleCleanup();
+  }
+
+  private scheduleCleanup() {
+    new TimedTrigger(
+      (timerFired) => {
+        const fullCleanup = !timerFired;  // We are in the idle period and should have nothing on the stack
+        this.cleanup(fullCleanup);
+        this.scheduleCleanup();
+      },
+      CLEANUP_PERIOD,
+      true // use idle callback when possible. 
+    );
+  }
+
+  public cleanup(fullCleanup: boolean) {
+    if (fullCleanup) {
+      // We are in the idle period and should have nothing on the stack
+      if (this.stackSize() > 0) {
+        this._flowletStack = [];
```

Comment:
Should we log when this happens so we know there is cases where we completely empty the stack?

Reply:
I have to expose some form of logging api to do this. I'll add that.